avocado/plugins/spawners/lxc.py

```diff
@@ -0,0 +1,200 @@
+from avocado.core.plugin_interfaces import Init, Spawner
+from avocado.core.settings import settings
+from avocado.core.spawners.common import SpawnerMixin, SpawnMethod
+
+
+class LXCStreamsFile:
+
+    def __init__(self):
+        self.fd = None
+        self.path = None
+
+    def fileno(self):
+        return self.fd
+
+    def read(self):
+        with open(self.path, "r") as fp:
+            return fp.read()
+
+    def __enter__(self):
+        self.fd, self.path = tempfile.mkstemp()
+        return self
+
+    def __exit__(self, *args):
+        os.remove(self.path)
+
+
+class LXCSpawnerInit(Init):
+
+    description = 'LXC (container) based spawner initialization'
+
+    def initialize(self):
+        section = 'spawner.lxc'
+
+        help_msg = 'Distro for the LXC container'
+        settings.register_option(
+            section=section,
+            key='dist',
+            help_msg=help_msg,
+            default='fedora')
+
+        help_msg = 'Release of the LXC container (depends on the choice of distro)'
+        settings.register_option(
+            section=section,
+            key='release',
+            help_msg=help_msg,
+            default='32')
+
+        help_msg = 'Architecture of the LXC container'
+        settings.register_option(
+            section=section,
+            key='arch',
+            help_msg=help_msg,
+            default='i386')
+
+
+class LXCSpawner(Spawner, SpawnerMixin):
+
+    description = 'LXC (container) based spawner'
+    METHODS = [SpawnMethod.STANDALONE_EXECUTABLE]
+
+    @staticmethod
+    def run_container_cmd(container, command):
+        with LXCStreamsFile() as tmp_out, LXCStreamsFile() as tmp_err:
+            exitcode = container.attach_wait(lxc.attach_run_command, command, stdout=tmp_out, stderr=tmp_err)
+            return exitcode, tmp_out.read(), tmp_err.read()
+
+    @staticmethod
+    async def run_container_cmd_async(container, command):
+        with LXCStreamsFile() as tmp_out, LXCStreamsFile() as tmp_err:
+            pid = container.attach(lxc.attach_run_command, command, stdout=tmp_out, stderr=tmp_err)
+            loop = asyncio.get_event_loop()
+            _, exitcode = await loop.run_in_executor(None, os.waitpid, pid, os.WUNTRACED)
+            return exitcode, tmp_out.read(), tmp_err.read()
+
+    @staticmethod
+    def is_task_alive(runtime_task):
+        if runtime_task.spawner_handle is None:
+            return False
+
+        if not LXC_AVAILABLE:
+            msg = 'LXC python bindings not available on the system'
+            runtime_task.status = msg
+            # we shouldn't reach this point
+            raise RuntimeError("LXC dependency is missing")
+
+        container = lxc.Container(runtime_task.spawner_handle)
+        if not container.defined:
+            return False
+        if not container.running:
+            return False
+
+        status, _, _ = LXCSpawner.run_container_cmd(container, ["pgrep", "avocado-runner"])
```

Comment:
I'm not sure if this pgrep is sufficient.  What happens if the process is in "zombie mode"? pgrep will return only the list of pids, but if for some reason the process is in <defunct> mode, this will return a false positive.

Reply:
Yes, this was just a minimal implementation for early testing and we definitely have to look at the corner cases. We could also try to filter out zombie or other wrong process states. Or would you suggest some other approach?